File: 2019/python/day4.py
pass_from = 264360
pass_to = 746325

MIN_LENGTH = 6


def criteria(p, mf):
    if len(p) != MIN_LENGTH or any(p[i] > p[i + 1] for i in range(MIN_LENGTH - 1)):
        return False
    groups = [p.count(ch) for ch in set(p)]
    return any([mf(g) for g in groups])


part1 = sum([criteria(str(p), lambda x: x >= 2) for p in range(pass_from, pass_to + 1)])
part2 = sum([criteria(str(p), lambda x: x == 2) for p in range(pass_from, pass_to + 1)])

print(part1, part2)

File: 2019/go/day4.go
package main

import (
	"fmt"
	"strconv"
)

const (
	passFrom  = 264360
	passTo    = 746325
	minLength = 6
)

func criteria(p string, mf func(int) bool) bool {
	if len(p) != minLength {
		return false
	}

	for i := 0; i < minLength-1; i++ {
		if p[i] > p[i+1] {
			return false
		}
	}

	charCounts := make(map[rune]int)
	for _, ch := range p {
		charCounts[ch]++
	}

	for _, count := range charCounts {
		if mf(count) {
			return true
		}
	}
	return false
}

func main() {
	part1 := 0
	part2 := 0

	for p := passFrom; p <= passTo; p++ {
		pStr := strconv.Itoa(p)

		if criteria(pStr, func(x int) bool { return x >= 2 }) {
			part1++
		}
		if criteria(pStr, func(x int) bool { return x == 2 }) {
			part2++
		}
	}

	fmt.Println(part1, part2)
}
